Update oregon.pdf to version 2/2023. Closes #1672 and #1642.

monocle opened the pull request:
* Create a new AcroFormMapper class to map the new PDF form fields to app variables.

* For existing forms, change the checkmark character for checkbox fields to "X".

@KentShikama The code isn't necessarily in a state that I would want to submit as a PR and I need to do more testing. But since the new form has to be used starting in March, I'm submitting it now to get the ball rolling.

Also, I'm just getting back into Python and have some dev tooling issues to figure out still and don't have auto formatting working.

KentShikama commented on src/backend/expungeservice/form_filling.py:
```diff
@@ -296,7 +306,7 @@ def _set_font_for_field(field, field_value, kid, warnings):
     def _build_font_string(field: PdfDict, field_value: str) -> Tuple[str, bool]:
         max_length = FormFilling._compute_field_max_length(field)
-        needs_shrink = len(field_value) > max_length
+        needs_shrink = len(field_value) > max_length if field_value != '/Yes' else False
```
What fields have /Yes?

monocle replied:
* Here are the files generated with John Common:
[OSP_Form_new.pdf](https://github.com/codeforpdx/recordexpungPDX/files/10832498/OSP_Form_new.pdf)
[COMMON A NAME_120000_baker_new.pdf](https://github.com/codeforpdx/recordexpungPDX/files/10832499/COMMON.A.NAME_120000_baker_new.pdf)
[COMMON NAME_110000_baker_new.pdf](https://github.com/codeforpdx/recordexpungPDX/files/10832500/COMMON.NAME_110000_baker_new.pdf)
[COMMON NAME_200000_benton_new.pdf](https://github.com/codeforpdx/recordexpungPDX/files/10832501/COMMON.NAME_200000_benton_new.pdf)

* For the checkbox fields, I had to use `/Yes` for the field values, but they generate the check mark symbol that Adobe programs in. I'm hoping that will fix the related Issue.

KentShikama commented on src/backend/expungeservice/form_filling.py:
```diff
@@ -236,17 +237,17 @@ def _build_pdf_for_eligible_case(
+            "has_class_a_misdemeanor": "X" if has_class_a_misdemeanor else "",
+            "has_class_bc_misdemeanor": "X" if has_class_bc_misdemeanor else "",
+            "has_violation_or_contempt_of_court": "X" if has_violation_or_contempt_of_court else "",
+            "has_probation_revoked": "X" if has_probation_revoked else "",
```
Did you check that the offsets work well for this.

monocle replied:
Here's how it looks with the "oregon_with_conviction.pdf"
<img width="759" alt="image" src="https://user-images.githubusercontent.com/120387970/221384407-ded7a588-0ace-47b3-8702-263c94279fdd.png">

[COMMON NAME_110000_baker_oregon_with_conviction.pdf](https://github.com/codeforpdx/recordexpungPDX/files/10832510/COMMON.NAME_110000_baker_oregon_with_conviction.pdf)

KentShikama replied:
Yeah are you able to create the box offset a bit? I had this issue too and offseting the box was a nice hack.

KentShikama reviewed (commented):
Do you have an example screenshot of the result on some test records or are you planning to test on dev?

monocle commented:
@KentShikama I've added the new `oregon.pdf` to the `oregon_with_arrest_order.pdf` and `oregon_with_convition_order.pdf` and updated the source code and tests.

Here are how the generated PDFs look:
* `oregon.pdf` 
[COMMON A NAME_120000_baker.pdf](https://github.com/codeforpdx/recordexpungPDX/files/10835164/COMMON.A.NAME_120000_baker.pdf)

* `oregon_with_arrest_order.pdf` [COMMON NAME_200000_benton_arrest.pdf](https://github.com/codeforpdx/recordexpungPDX/files/10835150/COMMON.NAME_200000_benton_arrest.pdf)
* `oregon_with_conviction_order.pdf`  [COMMON A NAME_120000_baker_conviction.pdf](https://github.com/codeforpdx/recordexpungPDX/files/10835151/COMMON.A.NAME_120000_baker_conviction.pdf)

Unfortunately, I now see that if I open these PDFs in Preview, I see the check marks. However, if on these Github links and the PDFs open in the Firefox browser, then the check marks are invisible.

Will keep working at it.

monocle commented:
Thanks @wittejm for helping out on this. What a life saver!

@KentShikama So in order to try to fix the inconsistent check mark issue, I tried going down the route for changing the annotations instead of the AcroForm.Fields. Using that, things seem to be working on Chrome, Firefox, Acrobat Reader and Safari and Preview. I still need to do more testing, but now out of nowhere I'm getting failing tests for `test_time_analyzer`.

I need to figure out what happened there. I don't remember touching anything related to that test file.

KentShikama commented:
Just some tests that didn't really scale through time. The dates are 2023, Feb 28 which happens to be today. Try bumping it to 2025 or something.

KentShikama commented:
@monocle Are you comfortable with handling the test fixes?

monocle commented:
@KentShikama It's not clear to me how to fix the time analyzer tests. Only one is failing today compared to four from yesterday. I just commented the test out for now if that's okay.

KentShikama commented:
Hmm. This could also be a leap year problem. 2023 isn't a leap year but 2020 and 2024 are and we are doing comparisons for X years before/after in this code a lot.

KentShikama commented:
Let's just keep the test in and merge this PR. I'm sure it will be "fixed" by tomorrow.

monocle commented:
@KentShikama I created an Issue to document the problem. I can check it tomorrow and if everything is passing, I can uncomment the test and submit a PR to put it back.

KentShikama commented:
@monocle Yup safe approach.

KentShikama commented on src/backend/tests/test_form_filling.py:
```diff
@@ -31,3 +35,351 @@ def test_normal_conviction_uses_multnomah_conviction_form():
+    #     pdf.write('foo_test')
+    #     assert font_string == "(/TimesNewRoman 6 Tf 0 g)"
+    #     assert len(pdf.warnings) == 1
+    #     assert pdf.warnings[0] == "The font size of \"123456789a123456\" was shrunk to fit the bounding box of \"DOB\". An addendum might be required if it still doesn't fit."
```
I prefer to not have commented code especially if the git history can preserve it. Was this intentionally left behind?

KentShikama replied:
Sounds good!

KentShikama commented on src/backend/expungeservice/form_filling.py:
```diff
@@ -256,23 +260,34 @@ def _build_pdf_for_eligible_case(
-                    annotation.update(PdfDict(AP=""))
-        pdf.Root.AcroForm.update(PdfDict(NeedAppearances=PdfObject("true")))
+
+        if "oregon" in pdf_path:
+            new_pdf = PDF(pdf_path, {"full_path": True})
+            new_pdf.update_annotations(form_data_dict)
+            warnings = new_pdf.warnings
+            pdf = new_pdf._pdf
+        else:
+            for field in pdf.Root.AcroForm.Fields:
+                form = from_dict(data_class=FormDataWithOrder, data=form_data_dict)
+                field_name = field.T.lower().replace(" ", "_").replace("(", "").replace(")", "")
+                field_value = getattr(form, field_name)
+                field.V = field_value
+                warnings += FormFilling._set_font(field, field.V)
+
+            # Since we are setting the values of the AcroForm.Fields, we need to
+            # remove the Appearance Dictionary ("/AP") of the PDF annotations in
+            # order for the value to appear in the PDF.
+            for page in pdf.pages:
+                annotations = page.get("/Annots")
+                if annotations:
+                    for annotation in annotations:
+                        annotation.update(PdfDict(AP=""))
+            pdf.Root.AcroForm.update(PdfDict(NeedAppearances=PdfObject("true")))
```
If I had to nitpick, there is two different levels of abstraction going on here. Most of the else could probably use something similar to your `update_annotations`. Just for the future. I don't want you to do another testing cycle here.

monocle replied:
Appreciate the feedback and agree. I hadn't looked at the Multnomah county forms until the last minute yesterday. I'm planning to unify the code more and hopefully I can just use the annotations approach with the Multnomah forms.

KentShikama commented on src/backend/expungeservice/form_filling.py:
```diff
@@ -178,8 +181,9 @@ def _build_pdf_for_case(
             return None
 
-    @staticmethod
+    @classmethod
```
Nitpick. I prefer to keep it staticmethod. Marking a function as a classmethod suggests to me that you're doing something more complicated with the cls but for just a constant, I think it adds more noise than helps. You can just have it top-level for the `CHECK_MARK` and just do `"has_conviction": CHECK_MARK if has_conviction else "",`

monocle replied:
Good to know. My Python is rusty so that was the first thing to come to mind when trying to pass that constant around.

KentShikama reviewed (approved):
Nice! And just on time for March.

KentShikama commented:
I'm going to go ahead and merge. We can continue discussing any comments if you want.
